docat/docat/utils.py

```diff
@@ -90,3 +97,230 @@ def calculate_token(password, salt):
+            )
+        )
+    )
+
+
+def get_project_details(upload_folder_path: Path, project_name: str):
+    """
+    Returns all versions and tags for a project.
+    """
+    docs_folder = upload_folder_path / project_name
+
+    if not docs_folder.exists():
+        return None
+
+    tags = [x for x in docs_folder.iterdir() if x.is_dir() and x.is_symlink()]
+
+    return ProjectDetailResponse(
+        name=project_name,
+        versions=sorted(
+            [
+                ProjectVersion(
+                    name=str(x.relative_to(docs_folder)),
+                    tags=[str(t.relative_to(docs_folder)) for t in tags if t.resolve() == x],
+                )
+                for x in docs_folder.iterdir()
+                if x.is_dir() and not x.is_symlink() and not (docs_folder / x.name / ".hidden").exists()
+            ],
+            key=lambda k: k.name,
+            reverse=True,
+        ),
+    )
+
+
+def index_all_projects(
+    upload_folder_path: Path,
+    index_db_path: Path,
+):
+    """
+    This will extract all content from all versions for each project,
+    and save it into index.json.
+    """
+    # drop existing index
+    if index_db_path.exists():
+        open(index_db_path, "w").close()
+
+    all_projects = get_all_projects(upload_folder_path).projects
+
+    for project in all_projects:
+        update_version_index_for_project(upload_folder_path, index_db_path, project)
+        update_file_index_for_project(upload_folder_path, index_db_path, project)
+
+
+def update_file_index_for_project(upload_folder_path: Path, index_db_path: Path, project: str):
+    """
+    Rebuilds the file index for all versions of the given project
+    """
+    index_db = TinyDB(index_db_path)
+    files_table = index_db.table("files")
+    files_table.remove(Query().project == project)
+
+    project_details = get_project_details(upload_folder_path, project)
+
+    for version in project_details.versions:
+        update_file_index_for_project_version(upload_folder_path, index_db_path, project, version.name)
+
+
+def update_file_index_for_project_version(upload_folder_path: Path, index_db_path: Path, project: str, version: str):
+    """
+    Removes existing indexes, and rebuilds it with the name of the contained files, and their content for html files.
+    """
+    docs_folder = upload_folder_path / project / version
+
+    if not docs_folder.exists():
+        return
+
+    remove_file_index_from_db(index_db_path, project, version)
+
+    for file in docs_folder.rglob("*"):
+        if not file.is_file():
+            continue
+
+        # save the file path
+        path = str(file.relative_to(docs_folder))
+        content = get_html_content_as_str(file) if file.name.endswith(".html") else ""
+
+        insert_file_index_into_db(index_db_path, project, version, path, content)
+
+
+def update_version_index_for_project(upload_folder_path: Path, index_db_path: Path, project: str):
```

Comment:
Why not unify both `update_version_index_for_project` and `update_file_index_for_project` into 
a single function `update_project_version_index` wich then call
`insert_version_into_version_index` and `update_file_index_for_project_version` ?

Reply:
I considered it but it made no sense because of tag function we provide. The way it is now, we can choose to just update the version index when we create a new tag.